// @adonisjs/fold/build/src/helpers.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const utils_1 = require("@poppinss/utils");
const toString = Function.prototype.toString;
function isEsm(value) {
    return value && value.__esModule;
}
exports.isEsm = isEsm;
function isPrimtiveConstructor(value) {
    return [String, Function, Object, Date, Number, Boolean].indexOf(value) > -1;
}
exports.isPrimtiveConstructor = isPrimtiveConstructor;
function isClass(fn) {
    return typeof (fn) === 'function' && /^class\s/.test(toString.call(fn));
}
exports.isClass = isClass;
function isObject(value) {
    return value && typeof (value) === 'object' && !Array.isArray(value);
}
exports.isObject = isObject;
function ensureIsFunction(callback, message) {
    if (typeof (callback) !== 'function') {
        throw new utils_1.Exception(message, 500, 'E_RUNTIME_EXCEPTION');
    }
}
exports.ensureIsFunction = ensureIsFunction;
function clearRequireCache(modulePath) {
    const cacheItem = require.cache[modulePath];
    if (!cacheItem) {
        return;
    }
    delete require.cache[modulePath];
    if (!cacheItem.parent) {
        return;
    }
    let i = cacheItem.parent.children.length;
    while (i--) {
        if (cacheItem.parent.children[i].id === modulePath) {
            cacheItem.parent.children.splice(i, 1);
        }
    }
}
exports.clearRequireCache = clearRequireCache;
