// @adonisjs/sink/src/base/KeyValueFile.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const BaseFile_1 = require("./BaseFile");
class KeyValueFile extends BaseFile_1.BaseFile {
    constructor(basePath) {
        super(basePath);
        this.$actions = [];
    }
    set(key, value) {
        this.$addAction('set', { key, value });
        return this;
    }
    unset(key) {
        this.$addAction('unset', { key });
        return this;
    }
    delete() {
        this.$addAction('delete');
        return this;
    }
    get(address, defaultValue) {
        return address ? this.filePointer.get(address, defaultValue) : this.filePointer.get();
    }
    exists() {
        return this.filePointer.exists();
    }
    commit() {
        this.$cdIn();
        const actions = this.$getCommitActions();
        const deleteFile = actions.find(({ action }) => action === 'delete');
        if (deleteFile) {
            this.filePointer.delete();
            this.$cdOut();
            return;
        }
        actions.forEach(({ action, body }) => {
            if (typeof (this[`on${action}`]) === 'function') {
                const handled = this[`on${action}`]('commit', body);
                if (handled) {
                    return;
                }
            }
            if (action === 'set') {
                this.filePointer.set(body.key, body.value);
                return;
            }
            if (action === 'unset') {
                this.filePointer.unset(body.key);
            }
        });
        this.filePointer.save();
        this.$cdOut();
    }
    rollback() {
        this.$cdIn();
        const actions = this.$getRevertActions();
        actions.forEach(({ action, body }) => {
            if (typeof (this[`on${action}`]) === 'function') {
                const handled = this[`on${action}`]('rollback', body);
                if (handled) {
                    return;
                }
            }
            if (action === 'set') {
                this.filePointer.unset(body.key);
                return;
            }
        });
        this.filePointer.save();
        this.$cdOut();
    }
}
exports.KeyValueFile = KeyValueFile;
