// @adonisjs/sink/index.js
"use strict";
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
const colors_1 = require("@poppinss/colors");
require("./src/disableLogger");
const pkg = __importStar(require("./package.json"));
const kleur = process.env.NODE_ENV === 'testing' ? new colors_1.FakeColors() : new colors_1.Colors();
exports.kleur = kleur;
function getPrompt() {
    const { Prompt, FakePrompt } = require('@poppinss/prompts');
    return process.env.NODE_ENV === 'testing' ? new FakePrompt() : new Prompt();
}
exports.getPrompt = getPrompt;
exports.sinkVersion = pkg.version;
var logger_1 = require("./src/logger");
exports.logger = logger_1.logger;
var RcFile_1 = require("./src/files/RcFile");
exports.RcFile = RcFile_1.RcFile;
var isEmptyDir_1 = require("./src/isEmptyDir");
exports.isEmptyDir = isEmptyDir_1.isEmptyDir;
var EnvFile_1 = require("./src/files/EnvFile");
exports.EnvFile = EnvFile_1.EnvFile;
var BaseFile_1 = require("./src/base/BaseFile");
exports.BaseFile = BaseFile_1.BaseFile;
var IniFile_1 = require("./src/formats/IniFile");
exports.IniFile = IniFile_1.IniFile;
var YamlFile_1 = require("./src/formats/YamlFile");
exports.YamlFile = YamlFile_1.YamlFile;
var JsonFile_1 = require("./src/formats/JsonFile");
exports.JsonFile = JsonFile_1.JsonFile;
var LinesFile_1 = require("./src/formats/LinesFile");
exports.LinesFile = LinesFile_1.LinesFile;
var copyTemplates_1 = require("./src/copyTemplates");
exports.copyTemplates = copyTemplates_1.copyTemplates;
var PackageFile_1 = require("./src/files/PackageFile");
exports.PackageFile = PackageFile_1.PackageFile;
var renderMarkdown_1 = require("./src/renderMarkdown");
exports.renderMarkdown = renderMarkdown_1.renderMarkdown;
var KeyValueFile_1 = require("./src/base/KeyValueFile");
exports.KeyValueFile = KeyValueFile_1.KeyValueFile;
var TemplateFile_1 = require("./src/formats/TemplateFile");
exports.TemplateFile = TemplateFile_1.TemplateFile;
var mrm_core_1 = require("mrm-core");
exports.copyFiles = mrm_core_1.copyFiles;
exports.deleteFiles = mrm_core_1.deleteFiles;
exports.makeDirs = mrm_core_1.makeDirs;
var executeInstructions_1 = require("./src/executeInstructions");
exports.executeInstructions = executeInstructions_1.executeInstructions;
