// @adonisjs/fold/build/src/Registrar/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const utils_1 = require("@poppinss/utils");
class Registrar {
    constructor(ioc, _basePath) {
        this.ioc = ioc;
        this._basePath = _basePath;
        this._providers = [];
        this._collected = false;
    }
    _loadProvider(providerPath) {
        providerPath = this._basePath ? utils_1.resolveFrom(this._basePath, providerPath) : providerPath;
        const provider = utils_1.esmRequire(providerPath);
        if (typeof (provider) !== 'function') {
            throw new Error(`Make sure export default the provider from ${providerPath}`);
        }
        return new provider(this.ioc);
    }
    _collect(providerPaths) {
        providerPaths.forEach((providerPath) => {
            const provider = this._loadProvider(providerPath);
            this._providers.push(provider);
            if (provider.provides) {
                this._collect(provider.provides);
            }
        });
    }
    useProviders(providersPaths) {
        this._providersPaths = providersPaths;
        return this;
    }
    register() {
        if (this._collected) {
            return this._providers;
        }
        this._collected = true;
        this._collect(this._providersPaths);
        this._providers.forEach((provider) => {
            if (typeof (provider.register) === 'function') {
                provider.register();
            }
        });
        return this._providers;
    }
    async boot() {
        const providers = this.register();
        for (let provider of providers) {
            if (typeof (provider.boot) === 'function') {
                await provider.boot();
            }
        }
    }
    async registerAndBoot() {
        const providers = this.register();
        await this.boot();
        return providers;
    }
}
exports.Registrar = Registrar;
