// @adonisjs/sink/src/formats/TemplateFile.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const mrm_core_1 = require("mrm-core");
const BaseFile_1 = require("../base/BaseFile");
class TemplateFile extends BaseFile_1.BaseFile {
    constructor(basePath, filename, templatePath) {
        super(basePath);
        this.$actions = [];
        this.removeOnRollback = true;
        this.overwrite = false;
        this.$cdIn();
        this.filePointer = mrm_core_1.template(filename, templatePath);
        this.$cdOut();
    }
    get() {
        return this.filePointer.get();
    }
    exists() {
        return this.filePointer.exists();
    }
    apply(contents) {
        this.filePointer.apply(contents);
        return this;
    }
    commit() {
        this.$cdIn();
        if (this.filePointer.exists() && !this.overwrite) {
            this.$cdOut();
            return;
        }
        this.filePointer.save();
        this.$cdOut();
    }
    rollback() {
        this.$cdIn();
        if (this.filePointer.exists() && this.removeOnRollback) {
            this.filePointer.delete();
        }
        this.$cdOut();
    }
}
exports.TemplateFile = TemplateFile;
